packages/dart_frog/lib/src/request.dart

```diff
@@ -115,11 +117,29 @@ class Request {
     return HttpMethod.values.firstWhere((m) => m.value == _request.method);
   }
 
+  Future<List<int>> _bytes() async {
+    if (_bodyBytesCompleter == null) {
+      _bodyBytesCompleter = Completer<List<int>>();
+      try {
+        final bytes = await _request.read().fold<List<int>>(
+          <int>[],
+          (previous, element) => previous..addAll(element),
+        );
+        _bodyBytesCompleter!.complete(bytes);
+      } catch (error, stackTrace) {
+        _bodyBytesCompleter!.completeError(error, stackTrace);
+      }
+    }
+    return _bodyBytesCompleter!.future;
+  }
+
   /// Returns a [Stream] representing the body.
-  Stream<List<int>> bytes() => _request.read();
+  Stream<List<int>> bytes() async* {
+    yield await _bytes();
+  }
 
   /// Returns a [Future] containing the body as a [String].
-  Future<String> body() => _request.readAsString();
+  Future<String> body() => utf8.decodeStream(bytes());
```

Comment:
you probably want to use body for json method or it will not get the cache

Reply:
yup good catch will address your comments in a few mins.